src/gflownet/models/seq_transformer.py

```diff
@@ -0,0 +1,137 @@
+            A batch of sequences.
+        cond: torch.Tensor
+            A tensor of conditional information.
+        batched: bool
+            If True, the it's assumed that the cond tensor is constant along a sequence, and the output is given
+            at each timestep (of the autoregressive process), which works because we are using causal self-attenion.
+            If False, only the last timesteps' output is returned, which one would use to sample the next token."""
+        x = self.embedding(xs.x)
+        x = self.pos(x)  # (time, batch, nemb)
+        x = self.encoder(x, src_key_padding_mask=xs.mask, mask=generate_square_subsequent_mask(x.shape[0]).to(x.device))
+        pooled_x = x[xs.lens - 1, torch.arange(x.shape[1])]  # (batch, nemb)
+
+        if self.use_cond:
+            cond_var = self.cond_embed(cond)  # (batch, nemb)
+            cond_var = torch.tile(cond_var, (x.shape[0], 1, 1)) if batched else cond_var
+            final_rep = torch.cat((x, cond_var), axis=-1) if batched else torch.cat((pooled_x, cond_var), axis=-1)
+        else:
+            final_rep = x if batched else pooled_x
+
+        out: torch.Tensor = self.output(final_rep)
+        ns = self.num_state_out
+        if batched:
+            # out is (time, batch, nout)
+            out = out.transpose(1, 0).contiguous().reshape((-1, out.shape[2]))  # (batch * time, nout)
+            # logit_idx tells us where (in the flattened array of outputs) the non-masked outputs are.
+            # E.g. if the batch is [["ABC", "VWXYZ"]], logit_idx would be [0, 1, 2, 5, 6, 7, 8, 9]
+            state_preds = out[xs.logit_idx, 0:ns]  # (proper_time, num_state_out)
+            stop_logits = out[xs.logit_idx, ns : ns + 1]  # (proper_time, 1)
+            add_node_logits = out[xs.logit_idx, ns + 1 :]  # (proper_time, nout - 1)
+            # `time` above is really max_time, whereas proper_time = sum(len(traj) for traj in xs))
+            # which is what we need to give to GraphActionCategorical
+        else:
+            # The default num_graphs is computed for the batched case, so we need to fix it here so that
+            # GraphActionCategorical knows how many "graphs" (sequence inputs) there are
+            xs.num_graphs = out.shape[0]
+            # out is (batch, nout)
+            state_preds = out[:, 0:ns]
+            stop_logits = out[:, ns : ns + 1]
+            add_node_logits = out[:, ns + 1 :]
+
+        return (
+            GraphActionCategorical(
+                xs,
+                logits=[stop_logits, add_node_logits],
+                keys=[None, None],
+                types=self.ctx.action_type_order,
+                slice_dict={},
+            ),
+            state_preds,
+        )
+
+
+def generate_square_subsequent_mask(sz: int):
+    """Generates an upper-triangular matrix of -inf, with zeros on diag."""
+    return torch.triu(torch.ones(sz, sz) * float("-inf"), diagonal=1)
+
+
+class PositionalEncoding(nn.Module):
```

Comment:
Consider newer (more efficient) PE

Reply:
I just reused Moksh's code here ^^"
Any suggestions on better implementations?

Reply:
If you care about very long sequences, then ALibi would make sense, otherwise, Rotary PE seems to work pretty well. See implementation of various PE here: https://github.com/lucidrains/x-transformers/blob/23f913aad5a93bff0a3c77267e04584a5f29f5ea/x_transformers/x_transformers.py#L215

Reply:
I added RotaryEmbedding, it looked interesting (but it would probably make sense to benchmark more of them at some point).